Show all submissions to a question in exam mode

rbessick5 opened the pull request:
Addresses #255 (and some portion of #29 )

I've left the extra assessment "pastsubmissions" within exampleCourse for testing (the other exams don't load because they include unsupported question types). Once everything checks out, I can make an additional commit to remove the unnecessary assessment.

If you like, I can also combine the two functions in userInstanceQuestionExam.js at lines 84-93 and 95-114, and remove the get_submission BLOCK (lines 11-19) from userInstanceQuestionExam.sql -- getting the most recent submission could be done while getting all the submissions.

mwest1066 commented on v2/pages/userInstanceQuestionExam/userInstanceQuestionExam.ejs:
```diff
@@ -47,6 +47,34 @@
+            
+            <% if (showAllSubmissions) { %>
+            <div class="panel-group">
+                <% for (sNum = 0; sNum < allSubmissions.length; sNum++) { 
```
Use https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach

rbessick5 replied:
I've made this change in 	d11c0ac

mwest1066 commented on v2/pages/userInstanceQuestionExam/userInstanceQuestionExam.ejs:
```diff
@@ -47,6 +47,34 @@
+                if (thisSubmission.correct === null) {
+                    panelType = "panel-info"; 
+                    submissionLabel = "<span class='label label-primary'>saved, not graded</span>"; 
+                }
```
use `} else {` (no linebreak)

mwest1066 commented on v2/pages/userInstanceQuestionExam/userInstanceQuestionExam.ejs:
```diff
@@ -47,6 +47,34 @@
+                    submissionLabel = "<span class='label label-primary'>saved, not graded</span>"; 
+                }
+                else if (thisSubmission.correct) {
+                    panelType = "panel-success";
```
I feel like having different panelTypes introduces too much visual noise in the display. I suggest sticking with just a single panelType (maybe `panel-info`?) and indicating correctness just with the label.

mwest1066 commented on v2/pages/userInstanceQuestionExam/userInstanceQuestionExam.ejs:
```diff
@@ -47,6 +47,34 @@
+                } %>
+                <div class="panel <%= panelType %> submission-block">
+                    <div class="panel-heading">
+                        <h3 class="panel-title">Submission <%= allSubmissions.length - sNum %> <%- submissionLabel %></h3>
```
Can we also add submission time here? Use the SQL functions `sprocs/format_date_*.sql` to format.

rbessick5 replied:
Added this in 	06330a8

mwest1066 commented on v2/pages/userInstanceQuestionExam/userInstanceQuestionExam.js:
```diff
@@ -90,6 +91,27 @@ router.get('/', function(req, res, next) {
+        // We can probably combine the previous function with the following one, and eliminate sql.get_submission
+        function(callback) {
+            res.locals.showAllSubmissions = false;
+            submissionFilePath = path.join(res.locals.course.path, "questions", res.locals.question.directory, "submission.html");
```
We shouldn't know anything about the internal structure of questions inside this file. Instead just unconditionally get all the past submission data and hand it off to the question renderer.

rbessick5 replied:
The showAllSubmissions flag will now be true if a submission exists; for problems which lack submission.html,, this will result in a list of submission panel-headers with no bodies, describing the submissions in order but not displaying them. Is this the behavior we want?

mwest1066 replied:
I think it's fine to display the fact that there were submissions even though we can't display the submission itself. Otherwise it would be confusing that some questions would appear to not have submissions and others would have them, which might lead people to think that the invisible submissions were not actually saved.

However, I think we should modify the question code (for `Calculation` questions now, at least) to render `Submission not shown` or similar for questions that don't have a `submission.html`.

mwest1066 commented on v2/pages/userInstanceQuestionExam/userInstanceQuestionExam.js:
```diff
@@ -94,23 +94,18 @@ router.get('/', function(req, res, next) {
+                if (ERR(err, callback)) return;
+                if (result.rowCount >= 1) {
+                    res.locals.showAllSubmissions = true;
+                    res.locals.allSubmissions = [];
```
Can just be `res.locals.allSubmissions = result.rows`

rbessick5 replied:
Fixed in 	92a812d

mwest1066 commented on v2/pages/userInstanceQuestionExam/userInstanceQuestionExam.ejs:
```diff
@@ -50,29 +50,23 @@
-                    submissionLabel = "<span class='label label-danger'>incorrect</span>";
-                } %>
-                <div class="panel <%= panelType %> submission-block">
+                <% allSubmissions.forEach(function (thisSubmission, sNum, allSubmissions) {
```
No need for the final `allSubmissions` argument to the `function()`. Also don't leave a space after `function`.

rbessick5 replied:
Fixed in 	06330a8

mwest1066 commented on v2/pages/userInstanceQuestionExam/userInstanceQuestionExam.ejs:
```diff
@@ -50,29 +50,23 @@
+                    } %>
+                <div class="panel panel-info submission-block">
                     <div class="panel-heading">
                         <h3 class="panel-title">Submission <%= allSubmissions.length - sNum %> <%- submissionLabel %></h3>
```
Add submission timestamp?

rbessick5 replied:
Added in 	06330a8

mwest1066 commented on v2/public/localscripts/calculationQuestion/SimpleClient.js:
```diff
@@ -297,7 +297,7 @@ define(["jquery", "underscore", "backbone", "rivets", "PrairieTemplate"], functi
             questionTemplate: "",
             answerTemplate: "",
-            submissionTemplate: "",
+            submissionTemplate: "Submission not shown.",
```
Also change this to `Saved answer`? Are there other places we that the student will see the word "submission"?

rbessick5 replied:
Fixed in 61dd42f

mwest1066 commented on v2/public/localscripts/calculationQuestion/SimpleClient.js:
```diff
@@ -248,11 +248,56 @@ define(["jquery", "underscore", "backbone", "rivets", "PrairieTemplate"], functi
+                this.rivetsView.unbind();
+            this.remove();
+        },
+    }); 
```
There is trailing whitespace on this line (extra spaces after the semicolon). Github doesn't show it, but it shows up in red when I do "git diff" on the command line (at least on my machine).

There are a few other trailing-whitespace errors in this PR. The general rule is that shouldn't be trailing whitespace on any line.

rbessick5 replied:
8624865 should get rid of all trailing whitespace and/or whitespace on blank lines

mwest1066 commented:
When I do Question 1 in HW1 for the exampleCourse (that is, `TPL 101` -> `HW1` -> `HW1.1. Addition of vectors in Cartesian coordinates`) and I submit an incorrect answer, then the true answer is not rendered, due to the following error on the browser console:

```
Uncaught TypeError: Cannot read property '0' of undefined
    at CalculationClient.renderSubmission (questionCalculation.js:77)
```

It looks like it's trying to render a submission history, but it doesn't exist? Is this just a problem because submission history isn't implemented yet for `Homework` assessments?

At the moment only a single `submission` is allowed for a given `variant` of a Homework question, but I think we will soon add the ability to allow multiple submissions per variant, even for homeworks. This will be useful for coding questions where you will want to repeatedly try the same variant if you have compile errors, for example. Anyway, this means that it would be good to support submission history rendering for Homeworks as well as Exams.

rbessick5 commented:
I see the problem here -- when I updated userInstanceQuestionExam.ejs to handle past submission rendering, I added a container with a list of "submission-body" objects, and put one submission into each (see /v2/public/localscripts/question.js line 20 ). I didn't change userInstanceQuestionHomework.ejs, but there was already such a block

rbessick5 commented:
Do you want to go ahead and implement past submission rendering for homework on this PR, or do we want to hold off?

rbessick5 commented:
The last commit adds submission rendering for homework questions; I don't have a test case for a homework problem allows multiple submissions, but this works correctly for the single submission model we currently have.

mwest1066 commented:
I get:

```
2016-11-12T21:25:14.851Z - error: Error page msg=/Users/mwest/git/PrairieLearn/v2/pages/userInstanceQuestionHomework/userInstanceQuestionHomework.ejs:80
    78|             <% } %>
    79|
 >> 80|             <% if (showAllSubmissions) { %>
    81|             <div class="panel-group">
    82|                 <% allSubmissions.forEach(function(thisSubmission, sNum) {
    83|                     if (thisSubmission.correct === null) {

showAllSubmissions is not defined, stack=ReferenceError: /Users/mwest/git/PrairieLearn/v2/pages/userInstanceQuestionHomework/userInstanceQuestionHomework.ejs:80
    78|             <% } %>
    79|
 >> 80|             <% if (showAllSubmissions) { %>
    81|             <div class="panel-group">
    82|                 <% allSubmissions.forEach(function(thisSubmission, sNum) {
    83|                     if (thisSubmission.correct === null) {

showAllSubmissions is not defined
    at eval (eval at <anonymous> (/Users/mwest/git/PrairieLearn/v2/node_modules/ejs/lib/ejs.js:485:12), <anonymous>:99:12)
    at returnedFn (/Users/mwest/git/PrairieLearn/v2/node_modules/ejs/lib/ejs.js:514:17)
    at View.exports.renderFile [as engine] (/Users/mwest/git/PrairieLearn/v2/node_modules/ejs/lib/ejs.js:358:31)
    at View.render (/Users/mwest/git/PrairieLearn/v2/node_modules/express/lib/view.js:126:8)
    at tryRender (/Users/mwest/git/PrairieLearn/v2/node_modules/express/lib/application.js:639:10)
    at EventEmitter.render (/Users/mwest/git/PrairieLearn/v2/node_modules/express/lib/application.js:591:3)
    at ServerResponse.render (/Users/mwest/git/PrairieLearn/v2/node_modules/express/lib/response.js:961:7)
    at /Users/mwest/git/PrairieLearn/v2/pages/userInstanceQuestionHomework/userInstanceQuestionHomework.js:214:13
    at /Users/mwest/git/PrairieLearn/v2/pages/userInstanceQuestionHomework/userInstanceQuestionHomework.js:192:9
    at /Users/mwest/git/PrairieLearn/v2/node_modules/async/lib/async.js:726:13, data=undefined
```

rbessick5 commented:
I'm not getting that error message on my set-up, so I'm not sure what might be preventing that from loading correctly. I store the value of showAllSubmissions in res.locals, where showTrueAnswer is located. Is the true answer block rendering on your set-up?

rbessick5 commented:
Got it -- it looks like my previous commit missed one of the files that needed to be updated (but, naturally, is updated on my local copy and so the code is working). This should (hopefully) eliminate the error you are seeing.
